server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -457,6 +461,38 @@ public QueryCachingPolicy getQueryCachingPolicy() {
         return cachingPolicy;
     }
 
+    /**
+     * A ref counter that can be used to delay primary relocation handoff via {@link #createRelocationDependency()}.
+     */
+    private final class RelocationCondition extends AbstractRefCounted {
+
+        private Runnable asyncActivation;
+
+        RelocationCondition(ShardRouting routing) {
+            super("relocation condition for [" + routing.shardId() + "][" + routing.allocationId() + "]");
+        }
+
+        @Override
+        protected void closeInternal() {
+            synchronized (this) {
+                if (asyncActivation != null) {
+                    threadPool.generic().execute(asyncActivation);
+                }
+            }
+        }
+
+        // Set the relocation context when receiving it and execute the handoff right away if no more conditions are waiting or create a
+        // Runnable to execute once all conditions have finished
+        void receivePrimaryContext(ReplicationTracker.PrimaryContext primaryContext, ActionListener<Void> listener) {
+            synchronized (this) {
+                if (decRef()) {
+                    doActivateWithPrimaryContext(primaryContext, listener);
```

Comment:
Could we do this outside the synchronized block? Not sure it is important, but simplifies reading the code.